### PR TITLE
Fix array input validation

### LINKED ISSUE #6128: Crash when adding validation messages to inputs within ArrayInputs
**What you were expecting:**
Since React 3.14.x, if an Inputs that lives inside ArrayInputs has an error message assigned it it, the app crashes on Save.

here's a video of this happening

https://user-images.githubusercontent.com/17087167/113691881-d5397100-96c4-11eb-9e7c-55eaf6b2102b.mov

I've tested this with both TextInputs and SelectInputs and they break, leading me to believe it's an issue with the `ArrayInput`. 

I've created a repro in [here](https://github.com/andrico1234/react-admin/tree/repro/nested-errors-in-arrays), you can run it and test the `PostEdit` page. 

This is the code causes the break

```tsx
import * as React from 'react';
import {
    ArrayInput,
    Edit,
    FormTab,
    SimpleFormIterator,
    TextInput,
    TabbedForm,
} from 'react-admin'; // eslint-disable-line import/no-unresolved

function validate(values: any) {
    const { authors } = values;
    const errors: any = {
        authors: [],
    };

    // works fine
    errors.singleRole = ['A role already exists with this author.'];

    // this doesn't works
    if (authors) {
        authors.forEach((author: any, i: number) => {
            if (!author) return;

            const error: any = {};

            if (author.role) {
                error.role = ['A role already exists with this author.'];
                errors.authors = [];
                errors.authors[i] = error;
            }
        });
    }

    return errors;
}

const PostEdit = ({ permissions, ...props }) => (
    <Edit {...props}>
        <TabbedForm warnWhenUnsavedChanges validate={validate}>
            <FormTab label="post.form.summary">
                {/* works fine */}
                <TextInput source="singleRole" />

                {/* doesn't work */}
                <ArrayInput source="authors">
                    <SimpleFormIterator>
                        <TextInput source="role" />
                    </SimpleFormIterator>
                </ArrayInput>
            </FormTab>
        </TabbedForm>
    </Edit>
);

export default PostEdit;

```

**Environment**

* React-admin version: 3.14.1+
* Last version that did not exhibit the issue (if applicable): 3.14.0
* React version: 17
* Browser: Chrome
* Stack trace (in case of a JS error):

```
Error: Objects are not valid as a React child (found: object with keys {role}). If you meant to render a collection of children, use an array instead.

    at ValidationError (http://localhost:8080/@fs/Users/andrico/Documents/Coding/react-admin/packages/ra-core/src/form/ValidationError.tsx:8:3)
    at InputHelperText (http://localhost:8080/@fs/Users/andrico/Documents/Coding/react-admin/packages/ra-ui-materialui/src/input/InputHelperText.tsx:8:3)
    at p
    at FormHelperText2 (http://localhost:8080/node_modules/.vite/chunk.HGLZTGBU.js?v=335e2996:61:5)
    at WithStyles(ForwardRef(FormHelperText2)) (http://localhost:8080/node_modules/.vite/chunk.POGTIVQJ.js?v=335e2996:362:31)
    at div
    at FormControl2 (http://localhost:8080/node_modules/.vite/chunk.4G2RE2EP.js?v=335e2996:73:5)
    at WithStyles(ForwardRef(FormControl2)) (http://localhost:8080/node_modules/.vite/chunk.POGTIVQJ.js?v=335e2996:362:31)
    at ArrayInput (http://localhost:8080/@fs/Users/andrico/Documents/Coding/react-admin/packages/ra-ui-materialui/src/input/ArrayInput.tsx:19:3)
    at div
    at FormInput (http://localhost:8080/@fs/Users/andrico/Documents/Coding/react-admin/packages/ra-ui-materialui/src/form/FormInput.tsx:30:5)
    at span
    at FormGroupContextProvider (http://localhost:8080/@fs/Users/andrico/Documents/Coding/react-admin/packages/ra-core/src/form/FormGroupContextProvider.tsx:10:3)
    at FormTab (http://localhost:8080/@fs/Users/andrico/Documents/Coding/react-admin/packages/ra-ui-materialui/src/form/FormTab.tsx:12:3)
    at Route2 (http://localhost:8080/node_modules/.vite/chunk.3HHKMOQI.js?v=335e2996:736:29)
    at div
    at form
    at TabbedFormView (http://localhost:8080/@fs/Users/andrico/Documents/Coding/react-admin/packages/ra-ui-materialui/src/form/TabbedFormView.tsx:24:5)
    at FormView (http://localhost:8080/@fs/Users/andrico/Documents/Coding/react-admin/packages/ra-core/src/form/FormWithRedirect.tsx:142:3)
    at ReactFinalForm (http://localhost:8080/node_modules/.vite/chunk.WW5ACNJ3.js?v=335e2996:124:22)
    at FormContextProvider (http://localhost:8080/@fs/Users/andrico/Documents/Coding/react-admin/packages/ra-core/src/form/FormContextProvider.tsx:5:3)
    at FormWithRedirect (http://localhost:8080/@fs/Users/andrico/Documents/Coding/react-admin/packages/ra-core/src/form/FormWithRedirect.tsx:25:3)
    at TabbedForm
    at div
    at Paper2 (http://localhost:8080/node_modules/.vite/chunk.3WWDLCXX.js?v=335e2996:52:5)
    at WithStyles(ForwardRef(Paper2)) (http://localhost:8080/node_modules/.vite/chunk.POGTIVQJ.js?v=335e2996:362:31)
    at Card2 (http://localhost:8080/node_modules/.vite/chunk.DGLX363H.js?v=335e2996:36:5)
    at WithStyles(ForwardRef(Card2)) (http://localhost:8080/node_modules/.vite/chunk.POGTIVQJ.js?v=335e2996:362:31)
    at div
    at div
```

### PR DESCRIPTION
**Pull request by djhi**
Fixes #6128
Supersedes #6133

image

image

### CODE REVIEW

File: examples/simple/src/posts/TagReferenceInput.tsx [hunk 1]
```diff
@@ -28,10 +34,18 @@ const TagReferenceInput = ({ ...props }) => {
         change('tags', []);
     };
 
+    const handleNewItem = (item: string) => {
+        console.log({ item });
```
Comment (fzaninotto) [inline examples/simple/src/posts/TagReferenceInput.tsx:38]: console.log() spotted

File: examples/simple/src/posts/TagReferenceInput.tsx [hunk 2]
```diff
@@ -28,10 +34,18 @@ const TagReferenceInput = ({ ...props }) => {
...
+        return new Promise((resolve, reject) => {});
+    };
+
     return (
         <div className={classes.input}>
             <ReferenceArrayInput {...props} filter={{ published: filter }}>
-                <AutocompleteArrayInput optionText="name.en" />
+                <AutocompleteArrayInput
+                    optionText="name.en"
+                    onNewItem={handleNewItem}
```
Comment (fzaninotto) [inline examples/simple/src/posts/TagReferenceInput.tsx:47]: this change seems unrelated to the current PR, am I right?

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx [hunk 3]
```diff
@@ -19,6 +19,10 @@ describe('<ReferenceArrayInputController />', () => {
         source: 'tag_ids',
     };
 
+    beforeEach(async () => {
+        await new Promise(resolve => setTimeout(resolve, 200));
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx:23]: why do you need that? Can't the delay be added after the tests that need it?

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx [hunk 4]
```diff
@@ -887,7 +898,7 @@ describe('<ReferenceArrayInputController />', () => {
 
     it('should call its children with the correct resource and basePath', () => {
         const children = jest.fn(() => null);
-        renderWithRedux(
+        const { unmount } = renderWithRedux(
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx:901]: linter warning

File: packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx [hunk 5]
```diff
@@ -905,4 +916,42 @@ describe('<ReferenceArrayInputController />', () => {
         expect(children.mock.calls[0][0].resource).toEqual('posts');
         expect(children.mock.calls[0][0].basePath).toEqual('/posts');
     });
+
+    it('should allow to refetch its data', async () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/input/ReferenceArrayInputController.spec.tsx:920]: same, this seems unrelated

File: packages/ra-ui-materialui/src/input/ArrayInput.spec.tsx [hunk 6]
```diff
@@ -126,4 +127,56 @@ describe('<ArrayInput />', () => {
             )
         ).toEqual(['bar', 'baz']);
     });
+
+    it('should apply validation to both itself and its inner inputs', async () => {
+        const { debug, getByText, getAllByLabelText, queryByText } = render(
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput.spec.tsx:132]: linter warning

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: examples/simple/src/posts/TagReferenceInput.tsx (hunk 2 above)
Comment (djhi) [reply examples/simple/src/posts/TagReferenceInput.tsx:47]: Yes, I messed up my commits. Working on too many branches :joy:

**Review by fzaninotto — APPROVED** [review]

**Pull request merged**